File: examples/tlplay-python/SettingsWidget.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the feather-tk project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl

import weakref

class CacheWidget(ftk.IWidget):

    def __init__(self, context, app, parent = None):
        ftk.IWidget.__init__(self, context, "CacheWidget", parent)
        
        self._app = weakref.ref(app)
        
        self._videoEdit = ftk.FloatEdit(context)
        self._videoEdit.range = ftk.RangeF(0.0, 128.0);
        self._videoEdit.step = 1.0
        self._videoEdit.largeStep = 10.0
        
        self._audioEdit = ftk.FloatEdit(context)
        self._audioEdit.range = ftk.RangeF(0.0, 128.0);
        self._audioEdit.step = 1.0
        self._audioEdit.largeStep = 10.0

        self._readBehindEdit = ftk.FloatEdit(context)
        self._readBehindEdit.range = ftk.RangeF(0.0, 2.0)

        self._layout = ftk.FormLayout(context, self)
        self._layout.spacingRole = ftk.SizeRole.SpacingSmall
        self._layout.addRow("Video cache (GB):", self._videoEdit)
        self._layout.addRow("Audio cache (GB):", self._audioEdit)
        self._layout.addRow("Read behind (seconds):", self._readBehindEdit)

        self._videoEdit.setCallback(self._videoCallback)
        self._audioEdit.setCallback(self._audioCallback)
        self._readBehindEdit.setCallback(self._readBehindCallback)

        self._cacheObserver = tl.timeline.ValueObserverPlayerCacheOptions(
            app.getSettingsModel().observeCache(), self._cacheUpdate)

    def setGeometry(self, value):
        ftk.IWidget.setGeometry(self, value)
        self._layout.setGeometry(value)
    
    def sizeHintEvent(self, event):
        self.setSizeHint(self._layout.sizeHint)

    def _videoCallback(self, value):
        if self._app:
            cache = self._app().getSettingsModel().observeCache().get()
            cache.videoGB = value
            self._app().getSettingsModel().setCache(cache)

    def _audioCallback(self, value):
        if self._app:
            cache = self._app().getSettingsModel().observeCache().get()
            cache.audioGB = value
            self._app().getSettingsModel().setCache(cache)

    def _readBehindCallback(self, value):
        if self._app:
            cache = self._app().getSettingsModel().observeCache().get()
            cache.readBehind = value
            self._app().getSettingsModel().setCache(cache)

    def _cacheUpdate(self, value):
        self._videoEdit.value = value.videoGB;
        self._audioEdit.value = value.audioGB;
        self._readBehindEdit.value = value.readBehind;

class Widget(ftk.IWidget):

    def __init__(self, context, app, parent = None):
        ftk.IWidget.__init__(self, context, "SettingsWidget", parent)
        
        layout = ftk.VerticalLayout(context)
        layout.marginRole = ftk.SizeRole.MarginSmall
        layout.spacingRole = ftk.SizeRole.SpacingSmall

        cacheGroupBox = ftk.GroupBox(context, "Cache", layout)
        self._cacheWidget = CacheWidget(context, app, cacheGroupBox)

        self._scrollWidget = ftk.ScrollWidget(context, ftk.ScrollType.Both, self)
        self._scrollWidget.border = False
        self._scrollWidget.widget = layout

    def setGeometry(self, value):
        ftk.IWidget.setGeometry(self, value)
        self._scrollWidget.setGeometry(value)
    
    def sizeHintEvent(self, event):
        self.setSizeHint(self._scrollWidget.sizeHint)

File: bin/tlplay/SettingsWidget.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2021-2024 Darby Johnston
// All rights reserved.

#include "SettingsWidget.h"

#include "App.h"
#include "SettingsModel.h"

#include <feather-tk/ui/GroupBox.h>

namespace tl
{
    namespace play
    {
        void CacheSettingsWidget::_init(
            const std::shared_ptr<feather_tk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<IWidget>& parent)
        {
            IWidget::_init(context, "CacheSettingsWidget", parent);

            _videoEdit = feather_tk::DoubleEdit::create(context);
            _videoEdit->setRange(0.0, 128.0);
            _videoEdit->setStep(1.0);
            _videoEdit->setLargeStep(10.0);

            _audioEdit = feather_tk::DoubleEdit::create(context);
            _audioEdit->setRange(0.0, 128.0);
            _audioEdit->setStep(1.0);
            _audioEdit->setLargeStep(10.0);

            _readBehindEdit = feather_tk::DoubleEdit::create(context);
            _readBehindEdit->setRange(0.0, 2.0);

            _layout = feather_tk::FormLayout::create(context, shared_from_this());
            _layout->setSpacingRole(feather_tk::SizeRole::SpacingSmall);
            _layout->addRow("Video cache (GB):", _videoEdit);
            _layout->addRow("Audio cache (GB):", _audioEdit);
            _layout->addRow("Read behind (seconds):", _readBehindEdit);

            std::weak_ptr<App> appWeak(app);
            _videoEdit->setCallback(
                [appWeak](double value)
                {
                    if (auto app = appWeak.lock())
                    {
                        auto cache = app->getSettingsModel()->getCache();
                        cache.videoGB = value;
                        app->getSettingsModel()->setCache(cache);
                    }
                });

            _audioEdit->setCallback(
                [appWeak](double value)
                {
                    if (auto app = appWeak.lock())
                    {
                        auto cache = app->getSettingsModel()->getCache();
                        cache.audioGB = value;
                        app->getSettingsModel()->setCache(cache);
                    }
                });

            _readBehindEdit->setCallback(
                [appWeak](double value)
                {
                    if (auto app = appWeak.lock())
                    {
                        auto cache = app->getSettingsModel()->getCache();
                        cache.readBehind = value;
                        app->getSettingsModel()->setCache(cache);
                    }
                });

            _cacheObserver = feather_tk::ValueObserver<timeline::PlayerCacheOptions>::create(
                app->getSettingsModel()->observeCache(),
                [this](const timeline::PlayerCacheOptions& value)
                {
                    _videoEdit->setValue(value.videoGB);
                    _audioEdit->setValue(value.audioGB);
                    _readBehindEdit->setValue(value.readBehind);
                });
        }

        CacheSettingsWidget::~CacheSettingsWidget()
        {
        }

        std::shared_ptr<CacheSettingsWidget> CacheSettingsWidget::create(
            const std::shared_ptr<feather_tk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<IWidget>& parent)
        {
            auto out = std::shared_ptr<CacheSettingsWidget>(new CacheSettingsWidget);
            out->_init(context, app, parent);
            return out;
        }

        void CacheSettingsWidget::setGeometry(const feather_tk::Box2I& value)
        {
            IWidget::setGeometry(value);
            _layout->setGeometry(value);
        }

        void CacheSettingsWidget::sizeHintEvent(const feather_tk::SizeHintEvent& event)
        {
            IWidget::sizeHintEvent(event);
            _setSizeHint(_layout->getSizeHint());
        }

        void SettingsWidget::_init(
            const std::shared_ptr<feather_tk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<IWidget>& parent)
        {
            IWidget::_init(context, "SettingsWidget", parent);
            _layout = feather_tk::VerticalLayout::create(context, shared_from_this());
            _layout->setMarginRole(feather_tk::SizeRole::MarginSmall);
            _layout->setSpacingRole(feather_tk::SizeRole::SpacingSmall);
            auto groupBox = feather_tk::GroupBox::create(context, "Cache", _layout);
            CacheSettingsWidget::create(context, app, groupBox);
        }

        SettingsWidget::~SettingsWidget()
        {
        }

        std::shared_ptr<SettingsWidget> SettingsWidget::create(
            const std::shared_ptr<feather_tk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<IWidget>& parent)
        {
            auto out = std::shared_ptr<SettingsWidget>(new SettingsWidget);
            out->_init(context, app, parent);
            return out;
        }

        void SettingsWidget::setGeometry(const feather_tk::Box2I& value)
        {
            IWidget::setGeometry(value);
            _layout->setGeometry(value);
        }

        void SettingsWidget::sizeHintEvent(const feather_tk::SizeHintEvent& event)
        {
            IWidget::sizeHintEvent(event);
            _setSizeHint(_layout->getSizeHint());
        }
    }
}